code/modules/reagents/chemistry_machinery/chem_simulator.dm

```diff
@@ -93,264 +92,286 @@
-
-		//List of all available properties
-		data["property_data_list"] = list()
-		for(var/datum/chem_property/P in GLOB.chemical_data.research_property_data)
-			data["property_codings"][P.name] = P.code
-			if(template_filter && !HAS_FLAG(P.category, template_filter))
-				continue
-			data["property_data_list"][P.name] = P.level
-			data["property_data_list"] = sortAssoc(data["property_data_list"])
-		//List of enabled properties
-		data["target_property_list"] = list()
-		for(var/datum/chem_property/P in creation_template)
-			data["target_property_list"][P.name] = P.level
-			if(template_filter && !HAS_FLAG(P.category, template_filter))
-				continue
-			//Override the editor level with the enabled property level
-			data["property_data_list"][P.name] = P.level
-
-		data["template_filter"] = list(
-				"MED" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_MEDICINE), PROPERTY_TYPE_MEDICINE),
-				"TOX" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_TOXICANT), PROPERTY_TYPE_TOXICANT),
-				"STI" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_STIMULANT), PROPERTY_TYPE_STIMULANT),
-				"REA" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_REACTANT), PROPERTY_TYPE_REACTANT),
-				"IRR" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_IRRITANT), PROPERTY_TYPE_IRRITANT),
-				"MET" = list(HAS_FLAG(template_filter, PROPERTY_TYPE_METABOLITE), PROPERTY_TYPE_METABOLITE)
-			)
+	tgui_interact(user)
 
-	else if(target && target.data && target.completed)
-		data["target_property_list"] = target.data.properties_to_assoc()
-		for(var/datum/chem_property/P in target.data.properties)
-			data["property_codings"][P.name] = P.code
-
-	if(reference && reference.data && reference.completed)
-		data["reference_property_list"] = reference.data.properties_to_assoc()
-		for(var/datum/chem_property/P in reference.data.properties)
-			data["property_codings"][P.name] = P.code
-
-	if(target_property)
-		data["target_property"] = target_property.name
-		data["target_info"] = target_property.description
-		data["target_categories"] = target_property.categories_to_string()
-	else
-		data["target_info"] = ""
-
-	if(reference_property)
-		data["reference_property"] = reference_property.name
-		data["reference_info"] = reference_property.description
-		data["reference_categories"] = reference_property.categories_to_string()
-	else
-		data["reference_info"] = ""
-
-	ui = SSnano.nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+/obj/structure/machinery/chem_simulator/tgui_interact(mob/user, datum/tgui/ui) //death to the chem simulator! All Hail the new chem simulator!
+	ui = SStgui.try_update_ui(user, src, ui)
 	if(!ui)
-		ui = new(user, src, ui_key, "chem_simulator.tmpl", "Synthesis Simulator", 800, 550)
-		ui.set_initial_data(data)
+		ui = new(user, src, "ChemSimulator", "Chemical Simulator")
 		ui.open()
 
-/obj/structure/machinery/chem_simulator/Topic(href, href_list)
+/obj/structure/machinery/chem_simulator/ui_data(mob/user)
```

Comment:
for reference, if auto update is on, this will be called every SStgui tick (which has a SS wait of 9) - feels like this proc is doing a decent amount of work, do you think it all needs to be in the auto updating proc? if something's only going to change after user interaction, you can put it in static data and during /ui_act, manually call update_static_data_for_all_viewers(). up to you, though!

Reply:
update_static_data_for_all_users()  completely swaps the ui to "please wait" one for solid 3 seconds while it uploads it again, and there isnt that many user interactions that wont make this unbearable to use. Almost all interactions the user will want to press something else shortly after the first one, and the only thing that is possible to move to static is the "known_properties", but then it would get annoyingly complicated to split that and the very dynamic part that checks and prevents users to add conflicting properties, in the end probably still leaving a solid chunk of code in this proc

Reply:
sounds good to me